ufl/matrix.py

```diff
@@ -0,0 +1,118 @@
+# -*- coding: utf-8 -*-
+"""This module defines the Matrix class."""
+
+# Copyright (C) 2021 India Marsden
+#
+# This file is part of UFL (https://www.fenicsproject.org)
+#
+# SPDX-License-Identifier:    LGPL-3.0-or-later
+
+
+from ufl.log import error
+from ufl.finiteelement import FiniteElementBase
+from ufl.domain import default_domain
+from ufl.form import BaseForm
+from ufl.argument import Argument
+from ufl.functionspace import AbstractFunctionSpace, FunctionSpace
+from ufl.utils.counted import counted_init
+
+
+# --- The Matrix class represents a matrix, an assembled two form ---
+
+class Matrix(BaseForm):
+    """UFL form argument type: Parent Representation of a form coefficient."""
+
+    __slots__ = (
+        "_count",
+        "_ufl_function_spaces",
+        "_repr",
+        "_hash",
+        "_ufl_shape",
+        "_arguments")
+    _globalcount = 0
+
+    def __getnewargs__(self):
+        return (self._ufl_function_spaces[0], self._ufl_function_spaces[1], self._count)
+
+    def __init__(self, row_space, column_space, count=None):
```

Comment:
bikeshed: should these be `test_space` and `trial_space` ?

Reply:
I think not because it will be possible to have a UFL Matrix resulting from things other than a variational form. Think about `interp`.